csrc/polymorphic_value.h

```diff
@@ -234,15 +234,38 @@ inline std::string toString(const PolymorphicValue& v) {
   return ss.str();
 }
 
+// NaNs are treated equal.
+template <typename T>
+inline bool isSameNanSensitive(const T& a, const T& b) {
+  if (std::isnan(a) && std::isnan(b)) {
+    return true;
+  }
+  return a == b;
+}
+
+template <typename T>
+inline bool isSameNanSensitive(
+    const std::complex<T>& a,
+    const std::complex<T>& b) {
+  return isSameNanSensitive(a.real(), b.real()) &&
```

Comment:
What if we have `nan + i`  and `nan - i`? Should we consider them as equal? We actually saw this issue in PyTorch eager mode, where one implementation of some math function returned `nan + i` while another returned `nan - i`, and there was a test asserting they are equal, and that test starts to become flaky at some point.

Reply:
Yeah, I expected that can of worms :) Even `nan==nan` is arguable and the consensus is actually not treating them equal. The real problem I wanted to solve is that `ExpressionEvaluator` runs `PrecomputedValues::validate` which triggers this equality check. FYI, I put the example in the test case. What do you think?

Reply:
Yeah, I understand what is the problem you are trying to solve, and I agree that this solves that problem. I am just concerned about an imaginary case like below:
```C++
Val* v = IrBuilder::create<Val>(std::complex<double>(nan, 1));
ExpressionEvaluator ee;
ee.bind(v, std::complex<double>(-1, nan));
```
With the implementation here, the above imaginary case will fail. But I believe this imaginary case should pass.

What is your opinion about this? I just feel that it makes more sense to implement this function like below:
```C++
  if ((std::isnan(a.real()) || std::isnan(a.imag())) && (std::isnan(b.real()) || std::isnan(b.imag()))) {
    return true;
  }
  return a == b;
```
but feel free to disagree with me. I don't think there is a widely accepted standard about how to compare complex nans.

Reply:
> ```
> Val* v = IrBuilder::create<Val>(std::complex<double>(nan, 1));
> ExpressionEvaluator ee;
> ee.bind(v, std::complex<double>(nan, -1));
> ```

Is there an actual use case for this? I expect users to bind the same value rather than sightly augmented values, so I'm inclined to say no for this case. However, backing up a step, why doesn't `ExpressionEvaluator` disallow binding precomputed values? If these values are precomputed (I assume by nvFuser), the user doesn't need to and probably shouldn't bind them to anything. Am I right?

Reply:
> Is there an actual use case for this? I expect users to bind the same value rather than sightly augmented values, so I'm inclined to say no for this case.

I don't have any concrete use case that we need to support. There is no consideration other than being more consistent with PyTorch:
```python
>>> nan_complex1 = torch.full((), complex(math.nan, 1))
>>> nan_complex2 = torch.full((), complex(-1, math.nan))
>>> torch.isclose(nan_complex1, nan_complex2, equal_nan=True)
tensor(True)
>>> torch.isclose(nan_complex1, nan_complex2)
tensor(False)
```
because I think `isSame` will not only be used with `ExpressionEvaluator`, but also serve as a general way to compare two `PolymorphicValue`s, and when facing a case that there is no clear answer, being consistent with PyTorch is always a good choice.

> However, backing up a step, why doesn't ExpressionEvaluator disallow binding precomputed values? If these values are precomputed (I assume by nvFuser), the user doesn't need to and probably shouldn't bind them to anything. Am I right?

My understanding is, sometimes it is just more convenient to allow binding a value twice, or allow binding a value to a constant `Val`. This convenience is mostly for shape inference. I don't think we are interested using `complex(-1, math.nan) == complex(math.nan, 1)` in shape inference.

Reply:
> when facing a case that there is no clear answer, being consistent with PyTorch is always a good choice.

That's a good point. I changed my implementation to match pytorch.